cpp/src/fil/infer.cu

```diff
@@ -531,45 +543,61 @@ template <int NITEMS, leaf_algo_t leaf_algo, bool cols_in_shmem,
 __global__ void infer_k(storage_type forest, predict_params params) {
   extern __shared__ char smem[];
   float* sdata = (float*)smem;
+  int sdata_stride = params.sdata_stride();
+  int rows_per_block = NITEMS << params.log2_threads_per_tree;
   int num_cols = params.num_cols;
-  for (size_t block_row0 = blockIdx.x * NITEMS; block_row0 < params.num_rows;
-       block_row0 += NITEMS * gridDim.x) {
-    size_t num_input_rows = min((size_t)NITEMS, params.num_rows - block_row0);
+  int thread_row0 = NITEMS * modpow2(threadIdx.x, params.log2_threads_per_tree);
+  for (size_t block_row0 = blockIdx.x * rows_per_block;
+       block_row0 < params.num_rows; block_row0 += rows_per_block * gridDim.x) {
+    int block_num_rows =
+      min((size_t)rows_per_block, params.num_rows - block_row0);
     const float* block_input = params.data + block_row0 * num_cols;
     if (cols_in_shmem) {
       // cache the row for all threads to reuse
-      size_t feature = 0;
 #pragma unroll
-      for (feature = threadIdx.x; feature < num_input_rows * num_cols;
-           feature += blockDim.x)
-        sdata[feature] = block_input[feature];
+      for (int input_idx = threadIdx.x; input_idx < block_num_rows * num_cols;
+           input_idx += blockDim.x) {
+        int row = input_idx / num_cols, col = input_idx % num_cols;
```

Comment:
Consider using division through multiplication here.

Reply:
It saves ~5% on the anomalous case of 1 tree of depth 1 and 4000 columns. So probably will save <1% in the real world, but might as well keep it.